modules/fedimint-wallet-server/src/lib.rs

```diff
@@ -1299,6 +1304,42 @@ impl Wallet {
         });
         (block_count_rx, fee_rate_rx)
     }
+
+    /// Shutdown the task group shared throughout fedimintd, giving 60 seconds
+    /// for other services to gracefully shutdown.
+    async fn graceful_shutdown(&self) {
+        if let Err(e) = self
+            .task_group
+            .clone()
+            .shutdown_join_all(Some(Duration::from_secs(60)))
+            .await
+        {
+            error!("Error while shutting down fedimintd task group: {e}");
+        }
+    }
+
+    /// If our bitcoin backend is stale, shutdown fedimintd. This is necessary
+    /// since we cannot participate in consensus if we're unable to observe the
+    /// chain.
+    ///
+    /// Our bitcoin backend is considered stale if:
+    ///   - it has not seen a block that our peers have observed
+    ///     `finality_delay` confirmations
+    ///   - it is unresponsive after 10 minutes
+    async fn shutdown_if_bitcoin_backend_is_stale(&self, new_consensus_block_count: u32) {
+        let strategy = backoff_util::custom_constant_backoff(Duration::from_secs(10), Some(6 * 10));
+        let get_block_count = || async { Ok(self.btc_rpc.get_block_count().await? as u32) };
```

Comment:
Ideally, we should not be doing rpcs in consensus processing. We should have our recent block count updated in the database / variable somewhere.